File: sms/src/main/java/dev/pitlor/sms/Messages.java
package dev.pitlor.sms;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import dev.pitlor.sms.repositories.MessageRepository;

public class Messages {
    private MessageRepository messageRepository;

    @Inject
    public Messages(MessageRepository messageRepository) {
        this.messageRepository = messageRepository;
    }

    public List<Message> readAllAfter(OffsetDateTime minimumTime) {
        List<List<String>> messageIds = messageRepository.getAllIdsAfter(minimumTime);

        List<Message> messages = new ArrayList<>();
        messages.addAll(messageIds.get(MessageRepository.SMS)
            .stream()
            .map(messageRepository::getSmsById)
            .map(Message::from)
            .collect(Collectors.toList()));
        messages.addAll(messageIds.get(MessageRepository.MMS)
            .stream()
            .map(messageRepository::getMmsById)
            .map(Message::from)
            .collect(Collectors.toList()));

        return messages;
    }
}

File: sms/src/main/java/dev/pitlor/sms/Messages.kt
package dev.pitlor.sms

import dev.pitlor.sms.repositories.MessageRepository
import java.time.OffsetDateTime
import javax.inject.Inject
import kotlin.coroutines.coroutineContext
import kotlin.reflect.KSuspendFunction2

class Messages @Inject constructor(private val messageRepository: MessageRepository) {
    fun readAllAfter(minimumTime: OffsetDateTime): List<Message> {
        val messageIds = messageRepository.getAllIdsAfter(minimumTime)
        return buildList(messageIds.mmsIds.size + messageIds.smsIds.size) {
            addAll(messageIds.smsIds
                .map(messageRepository::getSmsById)
                .map(Message::from)
            )
            addAll(messageIds.mmsIds
                .map(messageRepository::getMmsById)
                .map(Message::from)
            )
        }
    }

    suspend fun applyAll(saveMessages: suspend (Message) -> Unit) {
        messageRepository.applyAllMessages { messageDto ->
            val message =
                if (messageDto.isMms) messageRepository.getMmsById(messageDto.id)
                else messageRepository.getSmsById(messageDto.id)
            saveMessages(Message.from(message))
        }
    }
}